text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+```
+
+## Hidden, but coherent
+
+It is crucial to note that `Property for Thing` acts, with respect to coherence,
+as if it were public. That means that crate B, which is a reverse dependency of
+A, may not write:
+
+```rust
+// crate B:
+
+impl Property for Thing { .. }
+```
+
+This implementation is rejected, by the compiler, as overlapping with the one
+specified in crate A. The following is one *possible* error message,
+which is a variant of `E0119`, tailored for this specific case:
+
+```rust
+error[E0119]: conflicting implementations of trait `path::to::Property` for type `Thing`:
+ --> src/main.rs:<line>:<column>
+  |
+L | impl Property for Thing {
+  | ^^^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: conflicting implementation in crate `A`:
+          - crate impl Property for Thing
+    note: the implementation exists, but is hidden.
+```
+
+[unsound-named]: https://internals.rust-lang.org/t/looking-for-rfc-coauthors-on-named-impls/6275/2
+
+If the compiler did not reject this implementation, it would amount to the same
+problem as with named implementations, which would be decidedly unsound in Rust
+since [*"coherence is an invariant that unsafe code is allowed to assume"*][unsound-named].
+
+## Hiding things from yourself
+
+Thus far, we've only used `crate impl ..` to hide implementations.
+While we think that will be the most likely usage, we also believe
+that more fine grained control is beneficial for larger projects
+with many people involved in the development of those projects.
+Therefore, this RFC proposes that visibility modifiers (such as
+`pub(super)`) be permitted on `impl` to hide implementations from
+different parts of the crate.
+
+## `pub` by default
+
+Given that you are now allowed to put any visibility modifier on an `impl`,
+you might wonder if not specifying one is the same as private visibility
+(module visibility). This RFC specifies that it is not. Instead, default
+visibility is contextual. This means that `impl Foo for Bar { .. }` is
+equivalent to `pub impl Foo for Bar { .. }`. This is right for two reasons:
+
++ Changing the default visibility to module private would constitute a major
+  breaking change which would be too large to countenance even with an edition.
+
++ `pub` as a default for implementations is a *good* default, since it is what
+  you want most of the time.
+
+Having `pub` be permitted on `impl` also means that you can use the `vis`
+macro fragment specifier.
+
+To promote uniformity in the ecosystem, directly writing `pub impl ..` will
+be linted against with a lint named `redundant_vis` (or a naming in that spirit).
+A sample warning is:
+
+```rust
+warning: `pub` is redundant on `impl`
+ --> src/main.rs:<line>:<column>
+  |
+L |     pub impl Foo for Bar { .. }
+  |     ^^^ help: remove `pub`
+  |
+  = note: #[warn(redundant_vis)] on by default
+```
+
+## `priv` visibility
+
+To be able to make implementations module private, this RFC introduces a new
```

Comment:
Couldn't people who need this do `pub(in self)`? Seems like that would be less surprising than "priv"

Reply:
Or `priv` could be just made an alias of `pub(self)` (similar to `crate` being an alisa of `pub(crate)`).

Reply:
Apparently I was oblivious to the fact that `pub(self)` is already stable - so I'll make this section "possible future work".